mesop/components/autocomplete/autocomplete.ts

```diff
@@ -0,0 +1,168 @@
+        // https://github.com/google/mesop/issues/171
+        debounceTime(150),
+      )
+      .subscribe((event) => this.onInputDebounced(event));
+  }
+
+  ngOnInit(): void {
+    this.filteredOptions = this.autocompleteControl.valueChanges.pipe(
+      startWith(''),
+      map((value) => {
+        return this._filter(value || '', this._config.getOptionsList());
+      }),
+    );
+  }
+
+  ngOnDestroy(): void {
+    this.inputSubject.unsubscribe();
+  }
+
+  ngOnChanges() {
+    this._config = AutocompleteType.deserializeBinary(
+      this.type.getValue() as unknown as Uint8Array,
+    );
+  }
+
+  config(): AutocompleteType {
+    return this._config;
+  }
+  getStyle(): string {
+    return formatStyle(this.style);
+  }
+
+  getColor(): 'primary' | 'accent' | 'warn' {
+    return this.config().getColor() as 'primary' | 'accent' | 'warn';
+  }
+
+  getFloatLabel(): 'always' | 'auto' {
+    return this.config().getFloatLabel() as 'always' | 'auto';
+  }
+
+  getAppearance(): 'fill' | 'outline' {
+    return this.config().getAppearance() as 'fill' | 'outline';
+  }
+
+  getSubscriptSizing(): 'fixed' | 'dynamic' {
+    return this.config().getSubscriptSizing() as 'fixed' | 'dynamic';
+  }
+
+  onInput(event: Event): void {
+    this.inputSubject.next(event);
+  }
+
+  onInputDebounced(event: Event): void {
+    const userEvent = new UserEvent();
+    userEvent.setHandlerId(this.config().getOnInputHandlerId()!);
+    userEvent.setStringValue((event.target as HTMLInputElement).value);
+    userEvent.setKey(this.key);
+    this.channel.dispatch(userEvent);
+  }
+
+  onKeyUp(event: Event): void {
+    const keyboardEvent = event as KeyboardEvent;
+    if (keyboardEvent.key === 'Enter') {
+      const userEvent = new UserEvent();
+      userEvent.setHandlerId(this.config().getOnEnterHandlerId()!);
+      userEvent.setStringValue(this.autocompleteControl.value || '');
+      userEvent.setKey(this.key);
+      this.channel.dispatch(userEvent);
+    }
+  }
+
+  onAutocompleteSelectEvent(event: MatAutocompleteSelectedEvent): void {
+    const userEvent = new UserEvent();
+    userEvent.setHandlerId(this.config().getOnSelectionChangeHandlerId()!);
+    userEvent.setStringValue(this.autocompleteControl.value || '');
+    userEvent.setKey(this.key);
+    this.channel.dispatch(userEvent);
+  }
+
+  private _filter(
+    value: string,
+    options: AutocompleteOptionSet[],
+  ): AutocompleteOptionSet[] {
+    if (!value) {
+      return options;
+    }
+    const filterValue = value.toLowerCase();
+    const filteredOptions = new Array<AutocompleteOptionSet>();
+    for (const option of options) {
+      if (option.getOptionGroup()) {
+        const filteredOptionGroup = new AutocompleteOptionGroup();
+        filteredOptionGroup.setLabel(option.getOptionGroup()?.getLabel()!);
+        for (const subOption of option.getOptionGroup()?.getOptionsList()!) {
+          if (subOption?.getValue()?.toLowerCase().includes(filterValue)) {
```

Comment:
here and below - I think we should be filtering based on the label and not the value. I think the value is more of an internal representation (e.g. could be a serialized JSON), and the label is what the user sees, so we should filter based on the label.

Reply:
Doh. Good catch. Updated.